internal/graph/singleflight_resolver.go

```diff
@@ -0,0 +1,78 @@
+package graph
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/logger"
+)
+
+var (
+	deduplicatedDispatchesounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "deduplicated_dispatches_counter",
+		Help:      "The total number of calls to ResolveCheck that were deduplicated.",
+	})
+)
+
+type singleflightCheckResolver struct {
+	delegate CheckResolver
+	group    singleflight.Group
+	logger   logger.Logger
+}
+
+// SingleflightCheckResolverOpt defines an option that can be used to change the behavior of singleflightCheckResolver
+// instance.
+type SingleflightCheckResolverOpt func(*singleflightCheckResolver)
+
+func NewSingleflightCheckResolver(delegate CheckResolver, opts ...SingleflightCheckResolverOpt) *singleflightCheckResolver {
+	s := &singleflightCheckResolver{
+		delegate: delegate,
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	return s
+}
+
+// Close implements CheckResolver.
+func (s *singleflightCheckResolver) Close() {}
+
+// ResolveCheck implements CheckResolver.
+func (s *singleflightCheckResolver) ResolveCheck(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+) (*ResolveCheckResponse, error) {
+	key, err := CheckRequestCacheKey(req)
+	if err != nil {
+		s.logger.Error("singleflight cache key computation failed with error", zap.Error(err))
+		return nil, err
+	}
+
+	r, err, shared := s.group.Do(key, func() (interface{}, error) {
+		return s.delegate.ResolveCheck(ctx, req)
+	})
+
+	resp := r.(*ResolveCheckResponse)
+
+	if shared {
+		deduplicatedDispatchesounter.Inc()
```

Comment:
Keep in mind that `shared` is just a boolean, and thus it cannot distinguish how many de-duplications occurred from the single call. For example, image we de-duplicates the same request concurrently 1,000 times. The boolean would only count 2 not 1000.

I may actually recommend either:

a) extend the `singleflight`package so you also the get `count` of de-duplications. If you look at the internal code for that package that is available, but they don't expose it specifically.

or 

b) Write some code inspired by `singleflight` to achieve the same thing and report the count.

You may also consider looking at https://pkg.go.dev/resenje.org/singleflight#Group.Do, which provides a generic variant that is a little more friendly to the reader, and, more importantly, supports context cancellation as well.

Reply:
Please see the latest iteration. I think I do a better job of accurately measuring the deduplicated DB calls despite the limitations of the library.